horovod/common/basics.py

```diff
@@ -30,39 +32,79 @@ def __init__(self, pkg_path, *args):
         self.Sum = self.MPI_LIB_CTYPES.horovod_reduce_op_sum()
         self.Adasum = self.MPI_LIB_CTYPES.horovod_reduce_op_adasum()
 
-    def init(self, comm=None):
+    def init(self, comm=None, process_sets=None):
         """A function that initializes Horovod.
 
         Args:
-          comm: List specifying ranks for the communicator, relative to the MPI_COMM_WORLD
-            communicator OR the MPI communicator to use. Given communicator will be duplicated.
-            If None, Horovod will use MPI_COMM_WORLD Communicator.
+
+          comm: One of these possibilities:
+
+            1) List specifying ranks for the communicator, relative to the MPI_COMM_WORLD
+               communicator
+            2) None: Use all ranks of MPI_COMM_WORLD
+            3) MPI communicator to use. Given communicator will be duplicated.
+            4) List of MPI communicators. The first entry will be duplicated and used as
+               the global Horovod communicator. Horovod process sets will be built
+               corresponding to the remaining entries.
+
+          process_sets: If no explicit MPI communicators are passed, one of these possibilities:
+
+            1) None -- do not initialize any process sets
+            2) List[List[int]] -- initialize process sets containing these Horovod ranks
+            3) "dynamic": do not initialize any process sets now, but set environment
+               HOROVOD_DYNAMIC_PROCESS_SETS=1 .
         """
         if comm is None:
             comm = []
+        elif not isinstance(comm, list):
+            comm = [comm]
+        if process_sets is None:
+            process_sets = []
+        elif isinstance(process_sets, str) and process_sets.lower() == "dynamic":
+            process_sets = []
+            os.environ["HOROVOD_DYNAMIC_PROCESS_SETS"] = "1"
+
+        process_set_sizes = [len(ps) for ps in process_sets]
```

Comment:
Good point, thanks. 

I added two commits to ensure that each set of processes can only be added as a process set a single time, regardless of permutations. Also error handling in case there are mismatches in process set arguments to `hvd.init()` between calling processes is more thorough and friendlier now.